src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,134 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <backend.hpp>
+#include <Array.hpp>
+
+#include <af/array.h>
+#include <af/complex.h>
+#include <af/defines.h>
+#include <af/lapack.h>
+#include <arith.hpp>
+#include <blas.hpp>
+#include <cast.hpp>
+#include <common/ArrayInfo.hpp>
+#include <common/err_common.hpp>
+#include <diagonal.hpp>
+#include <handle.hpp>
+#include <logic.hpp>
+#include <math.hpp>
+#include <select.hpp>
+#include <svd.hpp>
+#include <transpose.hpp>
+
+using af::dim4;
+using namespace detail;
+
+template<typename T>
+Array<T> pinverse_svd(const Array<T> &in)
+{
+    // Moore-Penrose Pseudoinverse
+
+    in.eval();
+    int M = in.dims()[0];
+    int N = in.dims()[1];
+
+    // Compute SVD
+    typedef typename af::dtype_traits<T>::base_type Tr;
+    Array<Tr> sVec = createEmptyArray<Tr>(dim4(min(M, N)));
+    Array<T> u = createEmptyArray<T>(dim4(M, M));
+    Array<T> vT = createEmptyArray<T>(dim4(N, N));
+    svd<T, Tr>(sVec, u, vT, in);
+
+    Array<T> v = transpose(vT, true);
+
+    // Round down small values to zero to avoid large reciprocals later
+    Array<Tr> eps = createValueArray<Tr>(sVec.dims(), scalar<Tr>(1e-6));
```

Comment:
Implemented that relative tolerance calculation but epsilon will be 1e-6 by default instead of machine epsilon. There's an issue with SVD that prevents me from lowering this for float. I should probably write about that issue, but basically if I SVD a given array, make one of the singular values really small (1e-12 perhaps), multiply them back together, and SVD it again, the really small singular value isn't so small anymore (becomes something on the order of 1e-5). Thus I think our SVD cannot produce really small singular values and so I set the default tolerance that high.